java/arcs/core/allocator/Arc.kt

```diff
@@ -54,18 +58,17 @@ class Arc internal constructor(
      */
     var arcState: ArcState
         get() = arcStateInternal
-        private set(state) = sync(this) {
-            if (arcStateInternal != state) {
-                arcStateInternal = state
-                fireArcStateChange()
-            }
+        private set(state) {
+            arcStateInternal = state
         }
 
-    private fun onArcStateChange(handler: (ArcState) -> Unit) = sync(this) {
-        if (arcStateChangeHandlers.isEmpty()) {
-            registerChangeHandlerWithArcHosts()
+    private suspend fun onArcStateChange(handler: (ArcState) -> Unit) {
+        mutex.withLock {
+            if (arcStateChangeHandlers.isEmpty()) {
+                registerChangeHandlerWithArcHosts()
```

Comment:
I feel like this method is doing too much to justify calling it within the lock like this. Can any of it be split out (either by returning something that you can use outside of the `withLock` block, or just plain split into a separate method?)

Reply:
Let me rethink this. The general gist is to protect mutations of the handler callback list while iteration is happening, and to make recomputeArcState() safe. I think I can get rid of the mutate by using CopyOnWrite semantics (or CRDTs! :) )

Reply:
PTAL, switched to atomic-fu/copy-on-write